src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/Delegation/DesignTimeHelperResponseRewriter.cs

```diff
@@ -0,0 +1,115 @@
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Completion.Delegation
+{
+    internal class DesignTimeHelperResponseRewriter : DelegatedCompletionResponseRewriter
+    {
+        private static readonly IReadOnlyList<string> s_designTimeHelpers = new string[]
+        {
+            "__builder",
+            "__o",
+            "__RazorDirectiveTokenHelpers__",
+            "__tagHelperExecutionContext",
+            "__tagHelperRunner",
+            "__typeHelper",
+            "_Imports",
+            "BuildRenderTree"
+        };
+
+        private static readonly IReadOnlyList<CompletionItem> s_designTimeHelpersCompletionItems =
+            s_designTimeHelpers
+                .Select(item => new CompletionItem { Label = item })
+                .ToArray();
+
+        public override int Order => ExecutionBehaviorOrder.FiltersCompletionItems;
+
+        public override async Task<VSInternalCompletionList> RewriteAsync(
+            VSInternalCompletionList completionList,
+            int hostDocumentIndex,
+            DocumentContext hostDocumentContext,
+            DelegatedCompletionParams delegatedParameters,
+            CancellationToken cancellationToken)
+        {
+            if (delegatedParameters.ProjectedKind != RazorLanguageKind.CSharp)
+            {
+                return completionList;
+            }
+
+            var syntaxTree = await hostDocumentContext.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var owner = syntaxTree.GetOwner(hostDocumentIndex);
+            if (owner is null)
+            {
+                Debug.Fail("Owner should never be null.");
+                return completionList;
+            }
+
+            var sourceText = await hostDocumentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+
+            // We should remove Razor design time helpers from C#'s completion list. If the current identifier being targeted does not start with a double
+            // underscore, we trim out all items starting with "__" from the completion list. If the current identifier does start with a double underscore
+            // (e.g. "__ab[||]"), we only trim out common design time helpers from the completion list.
+
+            var filteredItems = completionList.Items.Except(s_designTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
+
+            if (ShouldRemoveAllDesignTimeItems(owner, sourceText))
+            {
+                filteredItems = filteredItems.Where(item => item.Label != null && !item.Label.StartsWith("__", StringComparison.Ordinal)).ToArray();
+            }
+
+            completionList.Items = filteredItems;
+            return completionList;
+        }
+
+        // If the current identifier starts with "__", only trim out common design time helpers from the list.
+        // In all other cases, trim out both common design time helpers and all completion items starting with "__".
```

Comment:
Is this just hedging against future compiler changes? Or is there some actual need for this?

Reply:
It's a good question. I originally was just copying from the COmpletionHandler but this does seem silly to have. I'll remove.